[CIR] Introduce CIR simplification

gitoleg opened the pull request:
This PR introduce cir simplification pass. The idea is to have a pass for the redundant operations removal/update.

Right now two pattern implemented, both related to the redundant `bool` operations. 
First pattern removes redundant casts from `bool` to `int` and back that for some reasons appear in the code. 
Second pattern removes sequential unary not operations (`!!`) .

For example, the code from the test is expanded from the simple check that is common for C code:

```
#define CHECK_PTR(ptr)  \
  do {                   \
    if (__builtin_expect((!!((ptr) == 0)), 0))\
      return -42; \
  } while(0)
```

I mark this PR as a draft for the following reasons:
1) I have no idea if it's useful for the community
2) There is a test fail - unfortunately current pattern rewriter run DCE underneath the hood and looks like we can't workaround it.
It's enough just to add an operation to the list - in this case `UnaryOp` - and call `applyOpPatternsAndFold`. I could rewrite a test a little in order to make everything non dead or implement a simple fix point algorithm for the particular task (I would do the former).

bcardosolopes commented:
Sorry I didn't had the time to look just yet, but should do that soon, can you update post rebase?

bcardosolopes commented on clang/lib/CIR/Dialect/Transforms/CIRSimplify.cpp:
```diff
@@ -0,0 +1,144 @@
+};
+
+class SimplifyBoolCasts : public mlir::OpRewritePattern<mlir::cir::CastOp> {
+
```
Any specific reason why can't this be implemented in `CastOp::fold`?

bcardosolopes commented on clang/lib/CIR/Dialect/Transforms/CIRSimplify.cpp:
```diff
@@ -0,0 +1,144 @@
+};
+
+class SimplifyUnaryNot : public mlir::OpRewritePattern<mlir::cir::UnaryOp> {
+
```
Similar question for UnaryOp, can't that be done with `fold` method?

bcardosolopes commented on clang/lib/CIR/Dialect/Transforms/CIRSimplify.cpp:
```diff
@@ -0,0 +1,144 @@
+};
+
+void CIRSimplifyPass::runOnOperation() {
+  RewritePatternSet patterns(&getContext());
```
If we need to create a new pass (after we go over the other comments), perhaps this is a good opportunity to rename MergeCleanups to CIRSimplify and add this over there instead.

gitoleg replied:
done!

bcardosolopes reviewed (changes requested):
> This PR introduce cir simplification pass. The idea is to have a pass for the redundant operations removal/update.
> 
> Right now two pattern implemented, both related to the redundant `bool` operations. First pattern removes redundant casts from `bool` to `int` and back that for some reasons appear in the code. Second pattern removes sequential unary not operations (`!!`) .

Neat!

> 1. I have no idea if it's useful for the community

Definetely great to get rid of these chains of redundant casts.

> 2. There is a test fail - unfortunately current pattern rewriter run DCE underneath the hood and looks like we can't workaround it.

Is this something related to the same issues we see with the canonicalizer? Is it worth adding this case to the bug tracking it?

> It's enough just to add an operation to the list - in this case `UnaryOp` - and call `applyOpPatternsAndFold`. I could rewrite a test a little in order to make everything non dead or implement a simple fix point algorithm for the particular task (I would do the former).

Former is a fine solution, I'm a bit worried about the overall approach though (more comments inline), I think we should try as much as possible to fix some of these problems at the MLIR level if necessary, I'm afraid those will keep piling up and might accumlate be too much technical debt in the future (e.g. CIR reinventing the wheel to workaround MLIR issues). Thoughts?

gitoleg commented:
@bcardosolopes  
Frankly, I don't know if it's possible to implement it as `fold`, since we need to erase operations and return values of different types - I tried and failed :(  For example, for the `CastOp::fold ` I returned  `%6`
```
%7 = cir.cast(bool_to_int, %6 : !cir.bool), !cir.int<s, 32>
%8 = cir.cast(int_to_bool, %7 : !cir.int<s, 32>), !cir.bool
```
but anyway got a type error. Maybe I'm doing something wrong though and didn't get how I should work wtih folders.

Another thought is that from the future optimizations point of view, I would have CIR simplified as much as we can do it, and better to do it explicit with the pass, than implicit with `fold` which is called somewhere. Well, maybe I'm wrong.

> Is this something related to the same issues we see with the canonicalizer? Is it worth adding this case to the bug tracking it?

I would say it's not a bug, it's a feature!!  which may hit us only in tests. It's slightly remind us the case with the canonicalizer since it also run the version of `applyPatternsAndFold` underneath of the hood. I don't think we need to create another issue here. 

To summarize:
1) I would create a separate pass to handle simplifications.  But I still need your advice in case I missed something with folders.
2) I fixed  tests  - just add "use" of operations.

bcardosolopes commented on clang/test/CIR/Transforms/simpl.cir:
```diff
@@ -0,0 +1,41 @@
+!s32i = !cir.int<s, 32>
+!s64i = !cir.int<s, 64>
+module {
+cir.func @foo(%arg0: !cir.ptr<!s32i>) {
```
What's the C code corresponding to this? Add a C code test instead with what you want to optimize.

For a `cir-opt` tests, better to split into two really simple CIR functions, each one testing the specific opt, so it's pretty clear you are adding coverage for different individual peepholes.

gitoleg replied:
done!

bcardosolopes commented on clang/test/CIR/CodeGen/unary.cpp:
```diff
@@ -155,28 +155,28 @@ int *inc_p(int *i) {
 // CHECK: cir.func @{{.+}}floats{{.+}}
-  +f; // CHECK: %{{[0-9]+}} = cir.unary(plus, %{{[0-9]+}}) : !cir.float, !cir.float
-  -f; // CHECK: %{{[0-9]+}} = cir.unary(minus, %{{[0-9]+}}) : !cir.float, !cir.float
+  f = +f; // CHECK: %{{[0-9]+}} = cir.unary(plus, %{{[0-9]+}}) : !cir.float, !cir.float
```
Do you need to change the source here because otherwise these operations are removed? Does everything else in the function touching `f` gets removed or just this one? Perhaps the function should return a float instead so that the value is used? If simplification is getting too aggresive, perhaps we should instead tie it up with `O1` or higher, because `-O0` shouldn't delete as much operations. Example: https://godbolt.org/z/6zP7qnjdP

bcardosolopes commented on clang/test/CIR/Transforms/simpl.cir:
```diff
@@ -0,0 +1,41 @@
+      %10 = cir.cast(int_to_bool, %9 : !s32i), !cir.bool
+      %11 = cir.unary(not, %10) : !cir.bool, !cir.bool
+      %12 = cir.cast(bool_to_int, %11 : !cir.bool), !s32i 
+      %13 = cir.cast(integral, %12 : !s32i), !s64i
```
The opt logic seems not to take into account integer promotion happening with casts here, which could be unsafe. Depending on the pattern you are trying to get rid off, it might be better to make sure CIRGen gets it right out of the bat.

bcardosolopes commented:
> For example, for the `CastOp::fold ` I returned `%6`
> 
> ```
> %7 = cir.cast(bool_to_int, %6 : !cir.bool), !cir.int<s, 32>
> %8 = cir.cast(int_to_bool, %7 : !cir.int<s, 32>), !cir.bool
> ```

From the canonicalization docs:

> ///  3. They can return an existing value or attribute that can be used instead
> ///     of the operation. The caller will remove the operation and use that
> ///     result instead.

In this case sounds like it wouldn't be a problem for the folder for `cir.cast(int_to_bool,` to return %6 directly? The folder would have to look at the defining op for `%7` and check it's a `bool_to_int`. Given the folder expects bool to be returned, looks like %6 would match?

> Another thought is that from the future optimizations point of view, I would have CIR simplified as much as we can do it, and better to do it explicit with the pass, than implicit with `fold` which is called somewhere. Well, maybe I'm wrong.

Simplifications might work better in face of canonicalized IR, but it's a blurry line sometimes where things should live. In the `cir.cast(int_to_bool,` case, it feels more like canonicalization. But overall if it's doing too much (deleting ops, blocks, adding new ops, etc), I agree with you. 

> 1. I would create a separate pass to handle simplifications.  But I still need your advice in case I missed something with folders.

MergeCleanups is not about `fold`ing, turns out that some of the easier things we moved into folders. I'd prefer we rename it to CIRSimplify and add transformations there as well.

My overall feedback is that I want to see the same optimizations you are trying to implement. But my feedback for this PR is that such simplification is doing too much in one place, pieces should be more independent, behave more like traditional 
 optimizations or canonicalization. For example: removing redundant casts is overall goodness that shouldn't be depending on optimizing unarys.

gitoleg commented:
@bcardosolopes 
Sorry for the delay! 

So. I rewrote everything  as folders for the correspondent operations and now it looks more elegant and readable than before. 
Briefly, what's underneath of the hood:
1) Implemented simplification of `!!` for boolean values 
2) Implemented simplification of redundant/unnecessary int/bool cast operations. 
3) Deleted my previous `CIRSimplify` pass and renamed `MergeCleanups` to `CIRSimplify`

The first one is clear. The second one may need some extra clarification.
Basically the approach is the same as it was earlier: collect all the subsequent casts operations, where each cast is one of the following kinds: `int_to_bool`, `bool_to_int` and `integral`. Next, check the head and the tail and do the folding in two cases, when head and tail are:
1) `bool_to_int`  and `int_to_bool`: the whole sequence of casts may be dropped as we had  boolean value before the first cast
2) `int_to_bool` and `int_to_bool`: the result of the first cast may be used, as no other cast (even the ext cast) won't change the
boolean result.


> Do you need to change the source here because otherwise these operations are removed? Does everything else in the function touching f gets removed or just this one?

Nope, only the results of unary operations that are not stored somewhere. And it's all because of `applyOpPatternsAndFold`, and `UnaryOp` that added to the operations to process: 
```
if (isa<BrOp, BrCondOp, ScopeOp, SwitchOp, 
            CastOp, TryOp, UnaryOp>(op))
    ops.push_back(op);
...

if (applyOpPatternsAndFold(ops, std::move(patterns)).failed())
    signalPassFailure();
```
So looks like `O0` and `O1` have nothing to do with it. Thus, if I just add `BinOp` to the list, the expression like `a+b` will also be removed once there is nothing on the left side.

> The opt logic seems not to take into account integer promotion happening with casts here, which could be unsafe. 

Well, the patterns I suggest have bool type as a result, so I would say it's safe to drop int promotion here: I explicitly check that the first cast kind is either `bool_to_int` (i.e. we already have boolean value before the first cast) or `int_to_bool` (i.e. boolean value is 
ready after the first cast). 

> Depending on the pattern you are trying to get rid off, it might be better to make sure CIRGen gets it right out of the bat.

Well, may be, but it's so easy to do it right here ...  The idea is to use just boolean value when we already got it, without any redundant casts.

bcardosolopes commented on clang/lib/CIR/Dialect/IR/CIRDialect.cpp:
```diff
@@ -513,7 +513,52 @@ LogicalResult CastOp::verify() {
   llvm_unreachable("Unknown CastOp kind?");
 }
 
+Value fold_casts(llvm::SmallVector<CastOp>& ops) {
```
These functions should follow the code style: https://llvm.github.io/clangir/GettingStarted/coding-guideline.html, e.g. `foldCasts`.

bcardosolopes commented on clang/lib/CIR/Dialect/IR/CIRDialect.cpp:
```diff
@@ -513,7 +513,52 @@ LogicalResult CastOp::verify() {
+  llvm::SmallVector<CastOp> casts;
+  collect_casts(*this, casts);
+  if (auto v = fold_casts(casts))
+    return v;
```
This should be unified with the switch below, is it really necessary to collect them? I feel if you handle each below the pass will run again until everything is covered, no need for us to do the extra work here.

gitoleg replied:
Let's say I have the next casts:
```
%1 = cir.cast(bool_to_int, %0 : !cir.bool), !s32i
%2 = cir.cast(integral, %1 : !s32i), !s64i
%3 = cir.cast(int_to_bool, %2 : !s64i), !cir.bool
```
The desired result is to return `%0` from `fold` for the cast `%3`, right? In this case `%3` will be removed, the number of users for `%2` will be equal to zero and the operation will be removed also, and the same will be done for `%1`, so all the casts will be eliminated.  
Now, if I try to do it just using folds, the only thing I could use is the result of `fold` for `%2` , but there is no evidence I can return some value when processing it.

So looks like I need to collect casts somehow. Is it really  possible via `fold` ? If it is, I will do it! But so far I don't see how to do it, sorry( 
Any thoughts?

bcardosolopes reviewed (changes requested):
Nice! Just some remaining comments/questions

bcardosolopes reviewed (changes requested):
I think the overall approach is good. Some extra thoughts on how to re-organize the approach a bit.

Ideally we should use one PR for each `::fold`'er, since they are not really dependent, just a quick note for next time around.

gitoleg commented on clang/test/CIR/Transforms/simpl.c:
```diff
@@ -0,0 +1,38 @@
+
+
+#define CHECK_PTR(ptr)  \
+  do {                   \
```
well, we don't have( maybe I will try to implement. From the other point of view, it kind of convenient to have `do .. while(0)` since we know where the macroses are) and may use this knowledge for some analysis.

bcardosolopes replied:
Good point. Right now we don't have a distinction between folders/simplifications meant for cleaning up versus optimization - MergeCleanups (now CIRSimplify) runs before all other passes (even before lifetime checker), which is still harmless because we don't do anything crazy. Looking fwd we might need to separate things that look more like canonicalization from optimization.

gitoleg commented on clang/lib/CIR/Dialect/IR/CIRDialect.cpp:
```diff
@@ -513,7 +513,52 @@ LogicalResult CastOp::verify() {
+  if (!isIntOrBoolCast(op))
+    return;
+
+  if (auto par = dyn_cast_or_null<CastOp>(op.getSrc().getDefiningOp()))
```
`par` means parent) looks like it's not a good variable name)

gitoleg commented on clang/lib/CIR/Dialect/IR/CIRDialect.cpp:
```diff
@@ -513,7 +513,52 @@ LogicalResult CastOp::verify() {
+    return;
+
+  if (auto par = dyn_cast_or_null<CastOp>(op.getSrc().getDefiningOp()))
+    collectCasts(par, ops);
```
got it!

gitoleg commented on clang/lib/CIR/Dialect/IR/CIRDialect.cpp:
```diff
@@ -535,6 +580,20 @@ OpFoldResult CastOp::fold(FoldAdaptor adaptor) {
+}
+
+OpFoldResult UnaryOp::fold(FoldAdaptor adaptor) {
+  if (isBoolNot(*this))
```
done!

gitoleg commented on clang/lib/CIR/Dialect/IR/CIRDialect.cpp:
```diff
@@ -513,7 +513,52 @@ LogicalResult CastOp::verify() {
 OpFoldResult CastOp::fold(FoldAdaptor adaptor) {
+  llvm::SmallVector<CastOp> casts;
+  collectCasts(*this, casts);
+  if (auto v = foldCasts(casts))
```
@bcardosolopes  so I implemented the chain folding as you said, but I'm not sure I understand how do you see  the call, i.e. where is a good place for it. Now I just added the call in the end of `CastOp::fold`. So, how do you see it?

bcardosolopes reviewed (approved):
LGTM thanks
